sophiread/FastSophiread/src/disk_io.cpp

```diff
@@ -435,3 +428,241 @@ void appendNeutronToHDF5(const std::string &out_file_name,
                          const std::vector<Neutron> &neutrons) {
   appendNeutronToHDF5(out_file_name, neutrons.cbegin(), neutrons.cend());
 }
+
+/**
+ * @brief Construct a new TPX3FileReader::TPX3FileReader object
+ *
+ * @param[in] filename
+ */
+TPX3FileReader::TPX3FileReader(const std::string &filename) {
+  fd = open(filename.c_str(), O_RDONLY);
+  if (fd == -1) {
+    spdlog::error("Failed to open file: {}", filename);
+    throw std::runtime_error("Failed to open file");
+  }
+
+  struct stat sb;
+  if (fstat(fd, &sb) == -1) {
+    spdlog::error("Failed to get file size");
+    close(fd);
+    throw std::runtime_error("Failed to get file size");
+  }
+
+  fileSize = sb.st_size;
+  map = static_cast<char *>(
+      mmap(nullptr, fileSize, PROT_READ, MAP_PRIVATE, fd, 0));
+  if (map == MAP_FAILED) {
+    spdlog::error("Failed to mmap file");
+    close(fd);
+    throw std::runtime_error("Failed to mmap file");
+  }
+
+  currentPosition = 0;
+  spdlog::info("Opened file: {}, size: {} bytes", filename, fileSize);
+}
+
+/**
+ * @brief Destroy the TPX3FileReader::TPX3FileReader object
+ */
+TPX3FileReader::~TPX3FileReader() {
+  if (map != MAP_FAILED) {
+    munmap(map, fileSize);
+  }
+  if (fd != -1) {
+    close(fd);
+  }
+}
+
+/**
+ * @brief Read a chunk of data from the file.
+ *
+ * @param[in] chunkSize: size of the chunk to read
+ */
+std::vector<char> TPX3FileReader::readChunk(size_t chunkSize) {
```

Comment:
Rather than returning by value which creates a copy, consider having the caller supply a `std::vector<char> &` as an output parameter. This allows for re-using the memory instead.